rust/core-lib/src/whitespace.rs

```diff
@@ -0,0 +1,143 @@
+// Copyright 2018 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+extern crate xi_rope;
+
+use xi_rope::Rope;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum Indentation {
+    Tabs,
+    Spaces(usize)
+}
+
+impl Indentation {
+    pub fn parse(rope: &Rope) -> Result<Option<Self>, String> {
+        let (lines, mut spaces, mut tabs) = (rope.lines_raw_all(), Vec::new(), false);
+
+        for line in lines {
+            match Indentation::parse_line(&line) {
+                Ok(Some(Indentation::Spaces(size))) => spaces.push(size),
+                Ok(Some(Indentation::Tabs)) => tabs = true,
+                Ok(None) => continue,
+                Err(e) => return Err(e),
+            }
+        }
+
+        match (tabs, spaces.len() > 0) {
+            (true, true) => return Err(String::from("Mixed indentation")),
+            (true, false) => Ok(Some(Indentation::Tabs)),
+            (false, true) => Ok(Some(Indentation::Spaces(extract_count(spaces)))),
+            _ => Ok(None)
+        }
+    }
+
+    /// Detects the indentation on a specific line.
+    /// Parses whitespace until first occurrence of something else
+    /// Tries to ignore lines which just contain whitespace, or comments, 
+    /// however, this is hard to implement effectively without a syntax tree
+    pub fn parse_line(line: &str) -> Result<Option<Self>, String> {
+        let (mut spaces, mut tabs) = (0, 0);
+        for char in line.chars() {
+            match char {
+                ' ' => spaces += 1,
+                '\t' => tabs += 1,
+                '/' | '#' | '\n' | '\r' => {
+                    spaces = 0;
+                    tabs = 0;
+                    break;
+                }
+                _ => break,
+            }
+        }
+
+        match (spaces > 0, tabs > 0) {
+            // Both spaces and tabs were found
+            (true, true) => Err(String::from("Mixed indentation")),
+            // Spaces were found
+            (true, false) => Ok(Some(Indentation::Spaces(spaces))),
+            // Tabs were found
+            (false, true) => Ok(Some(Indentation::Tabs)),
+            // Nothing was found
+            _ => Ok(None),
+        }
+    }
+}
+
+/// Uses a heuristic to calculate the amount of spaces used for indentation.
+fn extract_count(spaces: Vec<usize>) -> usize {
```

Comment:
I'm suspicious.

1. Many files are going to have a few lines that have unusual indentation; for instance, if you like to match lines up with open brackets, you may well have lines with an odd number of spaces, like

```rust
fn my_fun_func(&self,
               another_arg: usize) -> Fun {
    Fun::from(another_arg)
}
```
tl;dr: I would be very curious to actually see what this produces on a selection of source files. 

2. this is going to operate on every line, which seems unnecessary. In fact, having an entry in the vec for each line seems unnecessary.

Here's an alternative heuristic approach:

- Instead of a Vec, keep a `HashMap<usize, usize>` mapping the number of spaces to the number of times we see that number.
- Take the two most common counts; if one is the multiple of the other, our indentation unit is the smaller of the two.
- If the two most common counts do not represent > some constant fraction of the total, throw up our hands and fallback to the user's setting.

This is very handwavey and I haven't thought it through much; in particular I'm not sure if the 2 modal entries is enough, or if we might want to add a third. We'll also have trouble with files with few lines; we might have a minimum number of lines below which we aren't willing to guess?

I'm honestly curious how other editors approach this problem.